qualtran/bloqs/arithmetic/comparison.py

```diff
@@ -953,10 +954,61 @@ def wire_symbol(self, reg: Optional[Register], idx: Tuple[int, ...] = tuple()) -
             return TextBox(f"⨁(x = {self.val})")
         raise ValueError(f'Unknown register symbol {reg.name}')
 
+    def is_symbolic(self):
+        return is_symbolic(self.bitsize, self.val)
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: 'Soquet', target: 'Soquet'
+    ) -> Dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"cannot decompose symbolic {self}")
+
+        bits_k = x.reg.dtype.to_bits(self.val)
+
+        if self.bitsize == 1:
+            if self.val == 0:
+                x = bb.add(XGate(), q=x)
+            x, target = bb.add(CNOT(), ctrl=x, target=target)
+            if self.val == 0:
+                x = bb.add(XGate(), q=x)
+        elif self.bitsize == 2:
+            and_bloq = And(bits_k[0], bits_k[1])
+
+            xs = bb.split(x)
+            xs, and_xs = bb.add(and_bloq, ctrl=xs)
+            and_xs, target = bb.add(CNOT(), ctrl=and_xs, target=target)
+            xs = bb.add(and_bloq.adjoint(), ctrl=xs, target=and_xs)
+            x = bb.join(xs)
+        else:
+            multi_and_bloq = MultiAnd(tuple(bits_k))
+
+            xs = bb.split(x)
+            xs, junk, and_xs = bb.add(multi_and_bloq, ctrl=xs)
+            and_xs, target = bb.add(CNOT(), ctrl=and_xs, target=target)
+            xs = bb.add(multi_and_bloq.adjoint(), ctrl=xs, junk=junk, target=and_xs)
+            x = bb.join(xs)
+
+        return {'x': x, 'target': target}
+
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
-        # See: https://github.com/quantumlib/Qualtran/issues/219
-        # See: https://github.com/quantumlib/Qualtran/issues/217
-        return {(TGate(), 4 * (self.bitsize - 1))}
+        if self.is_symbolic():
+            op: Bloq
+            if not is_symbolic(self.bitsize) and self.bitsize <= 2:
+                if self.bitsize == 1:
+                    op = XGate()
+                else:
+                    cv = ssa.new_symbol('cv')
+                    op = And(cv, cv)
+            else:
+                op = MultiAnd(HasLength(self.bitsize))
+
+            bloq_counts: dict[Bloq, int] = defaultdict(lambda: 0)
+            bloq_counts[op] += 1
+            bloq_counts[op.adjoint()] += 1
+            bloq_counts[CNOT()] += 1
+            return set(bloq_counts.items())
+
+        return super().build_call_graph(ssa)
```

Comment:
nit: I generally prefer inverting this so the method isn't as indented

```python
if not self.is_symbolic():
  return super().build_call_graph

op = ...
return bloq_counts
```

That being said --- my original design for the call graph protocol was always to provide the counts you'd expect to find in e.g. a paper, which would use the `is_symbolic` path even for concrete numbers. Then you can reconcile them in e.g. the unit test by using the `generalizer` argument and built-in `generalizer` functions. Maybe I'll open an issue because I know this sort of thing has come up in other places

Reply:
Inverted the method.